metrics/metrics.go

```diff
@@ -0,0 +1,40 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	storageSystemMetrics map[string]prometheus.Collector
+	systemMapKey         string = "system_map"
+)
+
+func RegisterMetrics() {
+	if storageSystemMetrics == nil {
+		storageSystemMetrics = newStorageSystemMetrics()
+		for key := range storageSystemMetrics {
+			metrics.Registry.MustRegister(storageSystemMetrics[key])
+		}
+	}
+}
+
+// newStorageSystemMetrics returns a map of StorageSystem metrics
+func newStorageSystemMetrics() map[string]prometheus.Collector {
+	storageSystemMetrics := make(map[string]prometheus.Collector, 1)
+	storageSystemMetrics[systemMapKey] = newODFSystemMapMetrics()
+	return storageSystemMetrics
+}
+
+func newODFSystemMapMetrics() *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "odf",
+		Name:      systemMapKey,
+		Help:      "Map of ODF Storage System to Subsystems",
+	}, []string{"system", "subsystem", "kind"})
+}
+
+func ReportODFSystemMapMetrics(system, subsystem, kind string) {
+	metrics, _ := storageSystemMetrics[systemMapKey].(*prometheus.GaugeVec)
+	metrics.WithLabelValues(system, subsystem, kind).Set(1)
```

Comment:
I don't entirely understand what this function is doing. I'm assuming this is what records the actual mapping values, but what does `Set(1)` mean? I don't need this covered in the PR or comments, just a quick explanation here.

Reply:
It's setting the actual metric value (which can't be a map). The mapping itself is stored as labels of the metric.
And it's set to 1 so that Prometheus queries become simpler.